tests/component.test.js

```diff
@@ -0,0 +1,163 @@
+    let badXSpanComp = new Component('comp-name', 'comp-id', [l1, l2], -312423, 10, 'comp-entity', [p1, p2]);
+    let badYSpanComp = new Component('comp-name', 'comp-id', [l1, l2], 10, 0, 'comp-entity', [p1, p2]);
+    let badEntityComp = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, '', [p1, p2]);
+    let badPortComp = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, 'comp-entity', [new Port(), new Port()]);
+
+    expect(badComp.validate()).toBe(false);
+    expect(badNameComp.validate()).toBe(false);
+    expect(badIDComp.validate()).toBe(false);
+    expect(badLayerComp.validate()).toBe(false);
+    expect(badXSpanComp.validate()).toBe(false);
+    expect(badYSpanComp.validate()).toBe(false);
+    expect(badEntityComp.validate()).toBe(false);
+    expect(badPortComp.validate()).toBe(false);
+});
+
+test('validate Component: duplicate port labels', () => {
+    let c = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, 'comp-entity',
+            [new Port('port-name', l1, new Coord(5, 0)), new Port('port-name', l1, new Coord(0, 5))]);
+
+    expect(c.validate()).toBe(false);
+});
+
+test('validate Component: invalid port layers', () => {
+    let c = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, 'comp-entity',
+            [new Port('port-1', 'bad-layer', new Coord(5, 0)), new Port('port-2', l1, new Coord(0, 5))]);
+
+    expect(c.validate()).toBe(false);
+});
+
+test('validate Component: overlapping port locations', () => {
+    let c = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, 'comp-entity',
+            [new Port('port-1', l1, new Coord(5, 0)), new Port('port-2', l1, new Coord(5, 0))]);
+
+    expect(c.validate()).toBe(false);
+});
+
+test('validate Component: port locations', () => {
+    let c = new Component('comp-name', 'comp-id', [l1, l2], 10, 10, 'comp-entity',
+            [new Port('port-1', l1, new Coord())]);
+
+    for (let x = 0; x <= c.xSpan + 1; x++) {
+        for (let y = 0; y <= c.ySpan + 1; y++) {
+            c.ports[0].pos.setLocation(x, y);
+
+            if (x > c.xSpan || y > c.ySpan) {
+                // outside of the component is invalid
+                expect(c.validate()).toBe(false);
+            } else if ((x === 0 || x === c.xSpan) || (y === 0 || y === c.ySpan)) {
+                // edges and corners are valid
+                expect(c.validate()).toBe(true);
+            } else {
+                // anywhere else is invalid
+                expect(c.validate()).toBe(false);
+            }
+        }
+    }
+
+    // Spot checks just in case the above logic is off
+    c.ports[0].pos.setLocation(0, 0);
+    expect(c.validate()).toBe(true);
+
+    c.ports[0].pos.setLocation(4, 10);
+    expect(c.validate()).toBe(true);
+
+    c.ports[0].pos.setLocation(3, 6);
+    expect(c.validate()).toBe(false);
+
+    c.ports[0].pos.setLocation(0, 456);
+    expect(c.validate()).toBe(false);
```

Comment:
Yes, just do this instead of the above loops (you may also want to split these into their own tests)